Errors found by Coverity Scan static analysis

Hi, I have got some results of Coverity Scan static analysis. I filtered false positives and the following is the list of potential problems. I also tried to address them by PR. It may be not optimal and needs review (e.g. the explicit terminating nulls added after strncpy could be replaced by strlcpy, etc.).

<pre>
Error: VARARGS (CWE-237): [#def6]
ppp-2.4.7/chat/chat.c:503: va_init: Initializing va_list "args".
ppp-2.4.7/chat/chat.c:515: missing_va_end: va_end was not called for "args".
#  513|       if (to_stderr)
#  514|   	fprintf(stderr, "%s\n", line);
#  515|-> }
#  516|   
#  517|   /*

Error: VARARGS (CWE-237): [#def7]
ppp-2.4.7/chat/chat.c:526: va_init: Initializing va_list "args".
ppp-2.4.7/chat/chat.c:541: missing_va_end: va_end was not called for "args".
#  539|   	fprintf(stderr, "%s\n", line);
#  540|       terminate(code);
#  541|-> }
#  542|   
#  543|   int alarmed = 0;

Error: RESOURCE_LEAK (CWE-772): [#def14]
ppp-2.4.7/pppd/auth.c:461: alloc_fn: Storage is returned from allocation function "strdup".
ppp-2.4.7/pppd/auth.c:461: var_assign: Assigning: "fname" = storage returned from "strdup(*argv)".
ppp-2.4.7/pppd/auth.c:466: noescape: Resource "fname" is not freed or pointed-to in "option_error".
ppp-2.4.7/pppd/auth.c:467: leaked_storage: Variable "fname" going out of scope leaks the storage it points to.
#  465|       if (seteuid(getuid()) == -1) {
#  466|   	option_error("unable to reset uid before opening %s: %m", fname);
#  467|-> 	return 0;
#  468|       }
#  469|       ufile = fopen(fname, "re");

Error: RESOURCE_LEAK (CWE-772): [#def15]
ppp-2.4.7/pppd/auth.c:461: alloc_fn: Storage is returned from allocation function "strdup".
ppp-2.4.7/pppd/auth.c:461: var_assign: Assigning: "fname" = storage returned from "strdup(*argv)".
ppp-2.4.7/pppd/auth.c:469: noescape: Resource "fname" is not freed or pointed-to in "fopen".
ppp-2.4.7/pppd/auth.c:473: noescape: Resource "fname" is not freed or pointed-to in "option_error".
ppp-2.4.7/pppd/auth.c:474: leaked_storage: Variable "fname" going out of scope leaks the storage it points to.
#  472|       if (ufile == NULL) {
#  473|   	option_error("unable to open user login data file %s", fname);
#  474|-> 	return 0;
#  475|       }
#  476|       check_access(ufile, fname);

Error: BUFFER_SIZE_WARNING (CWE-120): [#def16]
ppp-2.4.7/pppd/eap-tls.c:695: buffer_size_warning: Calling strncpy with a maximum size argument of 1024 bytes on destination array "ets->peer" of size 1024 bytes might leave the destination string unterminated.
#  693|   	}
#  694|   
#  695|-> 	strncpy(ets->peer, esp->es_server.ea_peer, MAXWORDLEN);
#  696|   
#  697|   	dbglog( "getting eaptls secret" );

Error: BUFFER_SIZE_WARNING (CWE-120): [#def17]
ppp-2.4.7/pppd/eap-tls.c:783: buffer_size_warning: Calling strncpy with a maximum size argument of 1024 bytes on destination array "ets->peer" of size 1024 bytes might leave the destination string unterminated.
#  781|   	 */
#  782|   	if (esp->es_client.ea_peer)
#  783|-> 		strncpy(ets->peer, esp->es_client.ea_peer, MAXWORDLEN);
#  784|   	else
#  785|   		ets->peer[0] = 0;

Error: BUFFER_SIZE_WARNING (CWE-120): [#def18]
ppp-2.4.7/pppd/eap-tls.c:838: buffer_size_warning: Calling strncpy with a maximum size argument of 1024 bytes on destination array "ets->peercertfile" of size 1024 bytes might leave the destination string unterminated.
#  836|   	 */
#  837|   	if (servcertfile[0])
#  838|-> 		strncpy(ets->peercertfile, servcertfile, MAXWORDLEN);
#  839|   	else
#  840|   		ets->peercertfile[0] = 0;

Error: RESOURCE_LEAK (CWE-772): [#def20]
ppp-2.4.7/pppd/multilink.c:444: alloc_arg: "get_first_ethernet" allocates memory that is stored into "p".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:793:33: alloc_fn: Storage is returned from allocation function "strdup".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:793:33: var_assign: Assigning: "d" = "strdup(entry->d_name)".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:800:33: var_assign: Assigning: "eth_dev" = "d".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:820:9: var_assign: Assigning: "*_r" = "eth_dev".
ppp-2.4.7/pppd/multilink.c:445: noescape: Resource "p" is not freed or pointed-to in "get_if_hwaddr".
ppp-2.4.7/pppd/sys-linux.c:1866:35: noescape: "get_if_hwaddr(u_char *, char *)" does not free or save its parameter "name".
ppp-2.4.7/pppd/multilink.c:448: leaked_storage: Variable "p" going out of scope leaks the storage it points to.
#  446|   		ep->class = EPD_MAC;
#  447|   		ep->length = 6;
#  448|-> 		return 1;
#  449|   	}
#  450|   

Error: RESOURCE_LEAK (CWE-772): [#def21]
ppp-2.4.7/pppd/multilink.c:444: alloc_arg: "get_first_ethernet" allocates memory that is stored into "p".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:793:33: alloc_fn: Storage is returned from allocation function "strdup".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:793:33: var_assign: Assigning: "d" = "strdup(entry->d_name)".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:800:33: var_assign: Assigning: "eth_dev" = "d".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:820:9: var_assign: Assigning: "*_r" = "eth_dev".
ppp-2.4.7/pppd/multilink.c:445: noescape: Resource "p" is not freed or pointed-to in "get_if_hwaddr".
ppp-2.4.7/pppd/sys-linux.c:1866:35: noescape: "get_if_hwaddr(u_char *, char *)" does not free or save its parameter "name".
ppp-2.4.7/pppd/multilink.c:460: leaked_storage: Variable "p" going out of scope leaks the storage it points to.
#  458|   				ep->class = EPD_IP;
#  459|   				set_ip_epdisc(ep, addr);
#  460|-> 				return 1;
#  461|   			}
#  462|   		}

Error: RESOURCE_LEAK (CWE-772): [#def22]
ppp-2.4.7/pppd/multilink.c:444: alloc_arg: "get_first_ethernet" allocates memory that is stored into "p".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:793:33: alloc_fn: Storage is returned from allocation function "strdup".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:793:33: var_assign: Assigning: "d" = "strdup(entry->d_name)".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:800:33: var_assign: Assigning: "eth_dev" = "d".
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:820:9: var_assign: Assigning: "*_r" = "eth_dev".
ppp-2.4.7/pppd/multilink.c:445: noescape: Resource "p" is not freed or pointed-to in "get_if_hwaddr".
ppp-2.4.7/pppd/sys-linux.c:1866:35: noescape: "get_if_hwaddr(u_char *, char *)" does not free or save its parameter "name".
ppp-2.4.7/pppd/multilink.c:465: leaked_storage: Variable "p" going out of scope leaks the storage it points to.
#  463|   	}
#  464|   
#  465|-> 	return 0;
#  466|   }
#  467|   

Error: RESOURCE_LEAK (CWE-772): [#def27]
ppp-2.4.7/pppd/plugins/pppol2tp/openl2tp.c:188: alloc_fn: Storage is returned from allocation function "strdup".
ppp-2.4.7/pppd/plugins/pppol2tp/openl2tp.c:188: var_assign: Assigning: "user_name" = storage returned from "strdup(peer_authname)".
ppp-2.4.7/pppd/plugins/pppol2tp/openl2tp.c:228: noescape: Resource "user_name" is not freed or pointed-to in "strlen".
ppp-2.4.7/pppd/plugins/pppol2tp/openl2tp.c:229: noescape: Resource "user_name" is not freed or pointed-to in "memcpy". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/pppol2tp/openl2tp.c:249: leaked_storage: Variable "user_name" going out of scope leaks the storage it points to.
#  247|   	}
#  248|   
#  249|-> 	return;
#  250|   }
#  251|   

Error: BUFFER_SIZE_WARNING (CWE-120): [#def28]
ppp-2.4.7/pppd/plugins/radius/avpair.c:124: buffer_size_warning: Calling strncpy with a maximum size argument of 33 bytes on destination array "vp->name" of size 33 bytes might leave the destination string unterminated.
#  122|   							!= (VALUE_PAIR *) NULL)
#  123|   		{
#  124|-> 			strncpy (vp->name, pda->name, sizeof (vp->name));
#  125|   			vp->attribute = attrid;
#  126|   			vp->vendorcode = vendorcode;

Error: CPPCHECK_WARNING (CWE-401): [#def29]
ppp-2.4.7/pppd/plugins/radius/config.c:156: error[memleak]: Memory leak: iptr
#  154|   	else {
#  155|   		error("%s: auth_order: unknown keyword: %s", filename, p);
#  156|-> 		return (-1);
#  157|   	}
#  158|   

Error: CPPCHECK_WARNING (CWE-401): [#def30]
ppp-2.4.7/pppd/plugins/radius/config.c:168: error[memleak]: Memory leak: iptr
#  166|   		else {
#  167|   			error("%s: auth_order: unknown or unexpected keyword: %s", filename, p);
#  168|-> 			return (-1);
#  169|   		}
#  170|   	}

Error: COMPILER_WARNING: [#def31]
ppp-2.4.7/pppd/plugins/radius/config.c: scope_hint: In function 'rc_conf_str'
ppp-2.4.7/pppd/plugins/radius/config.c:273:2: warning: this 'if' clause does not guard... [-Wmisleading-indentation]
#  if (option == NULL)
#  ^~
ppp-2.4.7/pppd/plugins/radius/config.c:275:3: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the 'if'
#   return (char *)option->val;
#   ^~~~~~
#  271|   	option = find_option(optname, OT_STR);
#  272|   
#  273|-> 	if (option == NULL)
#  274|   		fatal("rc_conf_str: unkown config option requested: %s", optname);
#  275|   		return (char *)option->val;

Error: BUFFER_SIZE_WARNING (CWE-120): [#def32]
ppp-2.4.7/pppd/plugins/radius/radius.c:901: buffer_size_warning: Calling strncpy with a maximum size argument of 33 bytes on destination array "rstate.session_id" of size 33 bytes might leave the destination string unterminated.
#  899|       rstate.start_time = time(NULL);
#  900|   
#  901|->     strncpy(rstate.session_id, rc_mksid(), sizeof(rstate.session_id));
#  902|   
#  903|       rc_avpair_add(&send, PW_ACCT_SESSION_ID,

Error: CPPCHECK_WARNING (CWE-401): [#def33]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:71: error[memleak]: Memory leak: accts
#   69|       if ((fd = fopen(radrealms_config, "r")) == NULL) {
#   70|   	option_error("cannot open %s", radrealms_config);
#   71|-> 	return;
#   72|       } 
#   73|       info("Reading %s", radrealms_config);

Error: CPPCHECK_WARNING (CWE-401): [#def34]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:71: error[memleak]: Memory leak: auths
#   69|       if ((fd = fopen(radrealms_config, "r")) == NULL) {
#   70|   	option_error("cannot open %s", radrealms_config);
#   71|-> 	return;
#   72|       } 
#   73|       info("Reading %s", radrealms_config);

Error: RESOURCE_LEAK (CWE-772): [#def35]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: var_assign: Assigning: "accts" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:71: leaked_storage: Variable "accts" going out of scope leaks the storage it points to.
#   69|       if ((fd = fopen(radrealms_config, "r")) == NULL) {
#   70|   	option_error("cannot open %s", radrealms_config);
#   71|-> 	return;
#   72|       } 
#   73|       info("Reading %s", radrealms_config);

Error: RESOURCE_LEAK (CWE-772): [#def36]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: var_assign: Assigning: "auths" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:71: leaked_storage: Variable "auths" going out of scope leaks the storage it points to.
#   69|       if ((fd = fopen(radrealms_config, "r")) == NULL) {
#   70|   	option_error("cannot open %s", radrealms_config);
#   71|-> 	return;
#   72|       } 
#   73|       info("Reading %s", radrealms_config);

Error: RESOURCE_LEAK (CWE-772): [#def37]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: var_assign: Assigning: "accts" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:90: leaked_storage: Variable "accts" going out of scope leaks the storage it points to.
#   88|   	    option_error("%s: invalid line %d: %s", radrealms_config,
#   89|   			 line, buffer);
#   90|-> 	    return;
#   91|   	}
#   92|   	info("Parsing '%s' entry:", p);

Error: RESOURCE_LEAK (CWE-772): [#def38]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: var_assign: Assigning: "auths" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:90: leaked_storage: Variable "auths" going out of scope leaks the storage it points to.
#   88|   	    option_error("%s: invalid line %d: %s", radrealms_config,
#   89|   			 line, buffer);
#   90|-> 	    return;
#   91|   	}
#   92|   	info("Parsing '%s' entry:", p);

Error: RESOURCE_LEAK (CWE-772): [#def39]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: var_assign: Assigning: "accts" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:95: var_assign: Assigning: "s" = "accts".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:104: leaked_storage: Variable "s" going out of scope leaks the storage it points to.
ppp-2.4.7/pppd/plugins/radius/radrealms.c:104: leaked_storage: Variable "accts" going out of scope leaks the storage it points to.
#  102|   	    option_error("%s: realm name missing on line %d: %s",
#  103|   			 radrealms_config, line, buffer);
#  104|-> 	    return;
#  105|   	}
#  106|   

Error: RESOURCE_LEAK (CWE-772): [#def40]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: var_assign: Assigning: "auths" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:93: var_assign: Assigning: "s" = "auths".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:104: leaked_storage: Variable "auths" going out of scope leaks the storage it points to.
#  102|   	    option_error("%s: realm name missing on line %d: %s",
#  103|   			 radrealms_config, line, buffer);
#  104|-> 	    return;
#  105|   	}
#  106|   

Error: RESOURCE_LEAK (CWE-772): [#def41]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: var_assign: Assigning: "accts" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:95: var_assign: Assigning: "s" = "accts".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:114: leaked_storage: Variable "s" going out of scope leaks the storage it points to.
ppp-2.4.7/pppd/plugins/radius/radrealms.c:114: leaked_storage: Variable "accts" going out of scope leaks the storage it points to.
#  112|   		option_error("%s: server address missing on line %d: %s",
#  113|   			     radrealms_config, line, buffer);
#  114|-> 		return;
#  115|   	    }
#  116|   	    s->name[s->max] = strdup(p);

Error: RESOURCE_LEAK (CWE-772): [#def42]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: var_assign: Assigning: "auths" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:93: var_assign: Assigning: "s" = "auths".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:114: leaked_storage: Variable "auths" going out of scope leaks the storage it points to.
#  112|   		option_error("%s: server address missing on line %d: %s",
#  113|   			     radrealms_config, line, buffer);
#  114|-> 		return;
#  115|   	    }
#  116|   	    s->name[s->max] = strdup(p);

Error: RESOURCE_LEAK (CWE-772): [#def43]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:53: var_assign: Assigning: "accts" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:95: var_assign: Assigning: "s" = "accts".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:122: leaked_storage: Variable "s" going out of scope leaks the storage it points to.
ppp-2.4.7/pppd/plugins/radius/radrealms.c:122: leaked_storage: Variable "accts" going out of scope leaks the storage it points to.
#  120|   		option_error("%s: server port missing on line %d:  %s",
#  121|   			     radrealms_config, line, buffer);
#  122|-> 		return;
#  123|   	    }
#  124|   	    s->port[s->max] = atoi(p);

Error: RESOURCE_LEAK (CWE-772): [#def44]
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: alloc_fn: Storage is returned from allocation function "malloc".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:51: var_assign: Assigning: "auths" = storage returned from "malloc(88UL)".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:93: var_assign: Assigning: "s" = "auths".
ppp-2.4.7/pppd/plugins/radius/radrealms.c:122: leaked_storage: Variable "auths" going out of scope leaks the storage it points to.
#  120|   		option_error("%s: server port missing on line %d:  %s",
#  121|   			     radrealms_config, line, buffer);
#  122|-> 		return;
#  123|   	    }
#  124|   	    s->port[s->max] = atoi(p);

Error: COMPILER_WARNING: [#def47]
/usr/include/netdb.h:27: included_from: Included from here.
ppp-2.4.7/pppd/plugins/radius/includes.h:22: included_from: Included from here.
ppp-2.4.7/pppd/plugins/radius/sendserver.c:17: included_from: Included from here.
ppp-2.4.7/pppd/plugins/radius/sendserver.c: scope_hint: In function 'rc_send_server'
ppp-2.4.7/pppd/plugins/radius/sendserver.c:247:31: warning: implicit declaration of function 'rc_own_bind_ipaddress'; did you mean 'rc_own_ipaddress'? [-Wimplicit-function-declaration]
#  sin->sin_addr.s_addr = htonl(rc_own_bind_ipaddress());
#                               ^~~~~~~~~~~~~~~~~~~~~
#  245|   	memset ((char *) sin, '\0', (size_t) length);
#  246|   	sin->sin_family = AF_INET;
#  247|-> 	sin->sin_addr.s_addr = htonl(rc_own_bind_ipaddress());
#  248|   	sin->sin_port = htons ((unsigned short) 0);
#  249|   	if (bind (sockfd, (struct sockaddr *) sin, length) < 0 ||

Error: BUFFER_SIZE_WARNING (CWE-120): [#def48]
ppp-2.4.7/pppd/plugins/rp-pppoe/if.c:136: buffer_size_warning: Calling strncpy with a maximum size argument of 16 bytes on destination array "ifr.ifr_ifrn.ifrn_name" of size 16 bytes might leave the destination string unterminated.
#  134|       /* Fill in hardware address */
#  135|       if (hwaddr) {
#  136|-> 	strncpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
#  137|   	if (ioctl(fd, SIOCGIFHWADDR, &ifr) < 0) {
#  138|   	    error("Can't get hardware address for %s: %m", ifname);

Error: BUFFER_SIZE_WARNING (CWE-120): [#def49]
ppp-2.4.7/pppd/plugins/rp-pppoe/if.c:169: buffer_size_warning: Calling strncpy with a maximum size argument of 16 bytes on destination array "ifr.ifr_ifrn.ifrn_name" of size 16 bytes might leave the destination string unterminated.
#  167|       sa.sll_protocol = htons(type);
#  168|   
#  169|->     strncpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
#  170|       if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0) {
#  171|   	error("Could not get interface index for %s: %m", ifname);

Error: BUFFER_SIZE_WARNING (CWE-120): [#def50]
ppp-2.4.7/pppd/plugins/rp-pppoe/plugin.c:156: buffer_size_warning: Calling strncpy with a maximum size argument of 16 bytes on destination array "ifr.ifr_ifrn.ifrn_name" of size 16 bytes might leave the destination string unterminated.
#  154|   	goto errout;
#  155|       }
#  156|->     strncpy(ifr.ifr_name, conn->ifName, sizeof(ifr.ifr_name));
#  157|       if (ioctl(s, SIOCGIFMTU, &ifr) < 0) {
#  158|   	error("Can't get MTU for %s: %m", conn->ifName);

Error: BUFFER_SIZE_WARNING (CWE-120): [#def51]
ppp-2.4.7/pppd/plugins/rp-pppoe/plugin.c:329: buffer_size_warning: Calling strncpy with a maximum size argument of 16 bytes on destination array "ifr.ifr_ifrn.ifrn_name" of size 16 bytes might leave the destination string unterminated.
#  327|       /* Try getting interface index */
#  328|       if (r) {
#  329|-> 	strncpy(ifr.ifr_name, cmd, sizeof(ifr.ifr_name));
#  330|   	if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0) {
#  331|   	    r = 0;

Error: BUFFER_SIZE_WARNING (CWE-120): [#def52]
ppp-2.4.7/pppd/plugins/rp-pppoe/plugin.c:348: buffer_size_warning: Calling strncpy with a maximum size argument of 4096 bytes on destination array "devnam" of size 4096 bytes might leave the destination string unterminated.
#  346|       close(fd);
#  347|       if (r && doit) {
#  348|-> 	strncpy(devnam, cmd, sizeof(devnam));
#  349|   	if (the_channel != &pppoe_channel) {
#  350|   

Error: BUFFER_SIZE_WARNING (CWE-120): [#def53]
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:180: buffer_size_warning: Calling strncpy with a maximum size argument of 16 bytes on destination array "ifr.ifr_ifrn.ifrn_name" of size 16 bytes might leave the destination string unterminated.
#  178|       sa.sll_protocol = htons(type);
#  179|   
#  180|->     strncpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
#  181|       if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0) {
#  182|   	fatalSys("ioctl(SIOCFIGINDEX): Could not get interface index");

Error: COMPILER_WARNING: [#def54]
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:18: included_from: Included from here.
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c: scope_hint: In function 'sendPADI'
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe.h:299:2: warning: implicit declaration of function 'error'; did you mean 'perror'? [-Wimplicit-function-declaration]
#  error("Would create too-long packet"); \
#  ^~~~~
ppp-2.4.7/pppd/plugins/rp-pppoe/pppoe-discovery.c:460:5: note: in expansion of macro 'CHECK_ROOM'
#     CHECK_ROOM(cursor, packet.payload, plen);
#     ^~~~~~~~~~
#  297|   do {\
#  298|       if (((cursor)-(start))+(len) > MAX_PPPOE_PAYLOAD) { \
#  299|-> 	error("Would create too-long packet");	\
#  300|           return; \
#  301|       } \

Error: RESOURCE_LEAK (CWE-772): [#def55]
ppp-2.4.7/pppd/plugins/winbind.c:319: alloc_fn: Storage is returned from allocation function "fdopen".
ppp-2.4.7/pppd/plugins/winbind.c:319: var_assign: Assigning: "pipe_in" = storage returned from "fdopen(child_in[1], "w")".
ppp-2.4.7/pppd/plugins/winbind.c:327: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:333: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:339: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:345: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:350: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:357: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:367: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:377: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:381: noescape: Resource "pipe_in" is not freed or pointed-to in "fprintf". [Note: The source code implementation of the function has been overridden by a builtin model.]
ppp-2.4.7/pppd/plugins/winbind.c:382: noescape: Resource "pipe_in" is not freed or pointed-to in "fflush".
ppp-2.4.7/pppd/plugins/winbind.c:436: leaked_storage: Variable "pipe_in" going out of scope leaks the storage it points to.
#  434|           if (close(child_out[0]) == -1) {
#  435|                   notice("error closing pipe?!? for child OUT[0]");
#  436|->                 return NOT_AUTHENTICATED;
#  437|           }
#  438|   

Error: RESOURCE_LEAK (CWE-772): [#def63]
ppp-2.4.7/pppd/sys-linux.c:2172: open_fn: Returning handle opened by "socket".
ppp-2.4.7/pppd/sys-linux.c:2172: var_assign: Assigning: "s" = handle returned from "socket(2, 524290, 0)".
ppp-2.4.7/pppd/sys-linux.c:2177: noescape: Resource "s" is not freed or pointed-to in "ioctl".
ppp-2.4.7/pppd/sys-linux.c:2193: noescape: Resource "s" is not freed or pointed-to in "ioctl".
ppp-2.4.7/pppd/sys-linux.c:2249: leaked_handle: Handle variable "s" going out of scope leaks the handle.
# 2247|   	}
# 2248|       }
# 2249|->     return ok;
# 2250|   }
# 2251|   

Error: RESOURCE_LEAK (CWE-772): [#def64]
ppp-2.4.7/pppd/sys-linux.c:2713: open_fn: Returning handle opened by "open". [Note: The source code implementation of the function has been overridden by a user model.]
ppp-2.4.7/pppd/sys-linux.c:2713: var_assign: Assigning: "mfd" = handle returned from "open("/dev/ptmx", 524290)".
ppp-2.4.7/pppd/sys-linux.c:2716: noescape: Resource "mfd" is not freed or pointed-to in "ioctl".
ppp-2.4.7/pppd/sys-linux.c:2721: noescape: Resource "mfd" is not freed or pointed-to in "ioctl".
ppp-2.4.7/pppd/sys-linux.c:2735: overwrite_var: Overwriting handle "mfd" in "mfd = open(pty_name, 524290, 0)" leaks the handle.
# 2733|   	    slprintf(pty_name, sizeof(pty_name), "/dev/pty%c%x",
# 2734|   		     'p' + i / 16, i % 16);
# 2735|-> 	    mfd = open(pty_name, O_RDWR | O_CLOEXEC, 0);
# 2736|   	    if (mfd >= 0) {
# 2737|   		pty_name[5] = 't';

Error: BUFFER_SIZE_WARNING (CWE-120): [#def66]
ppp-2.4.7/pppstats/pppstats.c:153: buffer_size_warning: Calling strncpy with a maximum size argument of 16 bytes on destination array "req.b.ifr_ifrn.ifrn_name" of size 16 bytes might leave the destination string unterminated.
#  151|   #endif
#  152|   
#  153|->     strncpy(req.ifr_name, interface, sizeof(req.ifr_name));
#  154|       if (ioctl(s, SIOCGPPPSTATS, &req) < 0) {
#  155|   	fprintf(stderr, "%s: ", progname);

Error: BUFFER_SIZE_WARNING (CWE-120): [#def67]
ppp-2.4.7/pppstats/pppstats.c:179: buffer_size_warning: Calling strncpy with a maximum size argument of 16 bytes on destination array "creq.b.ifr_ifrn.ifrn_name" of size 16 bytes might leave the destination string unterminated.
#  177|   #endif
#  178|   
#  179|->     strncpy(creq.ifr_name, interface, sizeof(creq.ifr_name));
#  180|       if (ioctl(s, SIOCGPPPCSTATS, &creq) < 0) {
#  181|   	fprintf(stderr, "%s: ", progname);

Error: BUFFER_SIZE_WARNING (CWE-120): [#def68]
ppp-2.4.7/pppstats/pppstats.c:529: buffer_size_warning: Calling strncpy with a maximum size argument of 16 bytes on destination array "ifr.ifr_ifrn.ifrn_name" of size 16 bytes might leave the destination string unterminated.
#  527|   #define ifr_name ifr_ifrn.ifrn_name
#  528|   #endif
#  529|-> 	strncpy(ifr.ifr_name, interface, sizeof(ifr.ifr_name));
#  530|   	if (ioctl(s, SIOCGIFFLAGS, (caddr_t)&ifr) < 0) {
#  531|   	    fprintf(stderr, "%s: nonexistent interface '%s' specified\n",
</pre>

--- a/chat/chat.c
+++ b/chat/chat.c
@@ -512,6 +512,7 @@ void msgf __V((const char *fmt, ...))
 	syslog(LOG_INFO, "%s", line);
     if (to_stderr)
 	fprintf(stderr, "%s\n", line);
+    va_end(args);
 }
 
 /*
@@ -537,6 +538,7 @@ void fatal __V((int code, const char *fmt, ...))
 	syslog(LOG_ERR, "%s", line);
     if (to_stderr)
 	fprintf(stderr, "%s\n", line);
+    va_end(args);
     terminate(code);
 }
 

--- a/pppd/auth.c
+++ b/pppd/auth.c
@@ -430,6 +430,7 @@ setupapfile(argv)
     euid = geteuid();
     if (seteuid(getuid()) == -1) {
 	option_error("unable to reset uid before opening %s: %m", fname);
+        free(fname);
 	return 0;
     }
     ufile = fopen(fname, "r");
@@ -437,6 +438,7 @@ setupapfile(argv)
 	fatal("unable to regain privileges: %m");
     if (ufile == NULL) {
 	option_error("unable to open user login data file %s", fname);
+        free(fname);
 	return 0;
     }
     check_access(ufile, fname);
@@ -447,6 +449,7 @@ setupapfile(argv)
 	|| fgets(p, MAXSECRETLEN - 1, ufile) == NULL) {
 	fclose(ufile);
 	option_error("unable to read user login data file %s", fname);
+        free(fname);
 	return 0;
     }
     fclose(ufile);
@@ -468,6 +471,7 @@ setupapfile(argv)
 	explicit_passwd = 1;
     }
 
+    free(fname);
     return (1);
 }
 

--- a/pppd/multilink.c
+++ b/pppd/multilink.c
@@ -445,8 +445,12 @@ get_default_epdisc(ep)
 	if (p != 0 && get_if_hwaddr(ep->value, p) >= 0) {
 		ep->class = EPD_MAC;
 		ep->length = 6;
+		free(p);
 		return 1;
 	}
+
+	if (p)
+		free(p);
 
 	/* see if our hostname corresponds to a reasonable IP address */
 	hp = gethostbyname(hostname);

--- a/pppd/plugins/pppol2tp/openl2tp.c
+++ b/pppd/plugins/pppol2tp/openl2tp.c
@@ -246,6 +246,9 @@ out:
 		(*old_pppol2tp_ip_updown_hook)(tunnel_id, session_id, up);
 	}
 
+	if (user_name != NULL)
+		free(user_name);
+
 	return;
 }
 

--- a/pppd/plugins/radius/avpair.c
+++ b/pppd/plugins/radius/avpair.c
@@ -121,7 +121,7 @@ VALUE_PAIR *rc_avpair_new (int attrid, void *pval, int len, int vendorcode)
 		if ((vp = (VALUE_PAIR *) malloc (sizeof (VALUE_PAIR)))
 							!= (VALUE_PAIR *) NULL)
 		{
-			strncpy (vp->name, pda->name, sizeof (vp->name));
+			strlcpy (vp->name, pda->name, NAME_LENGTH);
 			vp->attribute = attrid;
 			vp->vendorcode = vendorcode;
 			vp->next = (VALUE_PAIR *) NULL;

--- a/pppd/plugins/radius/config.c
+++ b/pppd/plugins/radius/config.c
@@ -153,6 +153,7 @@ static int set_option_auo(char *filename, int line, OPTION *option, char *p)
 			*iptr = AUTH_RADIUS_FST;
 	else {
 		error("%s: auth_order: unknown keyword: %s", filename, p);
+		free(iptr);
 		return (-1);
 	}
 
@@ -165,6 +166,7 @@ static int set_option_auo(char *filename, int line, OPTION *option, char *p)
 			*iptr = (*iptr) | AUTH_RADIUS_SND;
 		else {
 			error("%s: auth_order: unknown or unexpected keyword: %s", filename, p);
+			free(iptr);
 			return (-1);
 		}
 	}
@@ -272,7 +274,7 @@ char *rc_conf_str(char *optname)
 
 	if (option == NULL)
 		fatal("rc_conf_str: unkown config option requested: %s", optname);
-		return (char *)option->val;
+	return (char *)option->val;
 }
 
 int rc_conf_int(char *optname)

--- a/pppd/plugins/radius/radius.c
+++ b/pppd/plugins/radius/radius.c
@@ -898,7 +898,7 @@ radius_acct_start(void)
 
     rstate.start_time = time(NULL);
 
-    strncpy(rstate.session_id, rc_mksid(), sizeof(rstate.session_id));
+    strlcpy(rstate.session_id, rc_mksid(), MAXSESSIONID);
 
     rc_avpair_add(&send, PW_ACCT_SESSION_ID,
 		   rstate.session_id, 0, VENDOR_NONE);

--- a/pppd/plugins/radius/radiusclient.h
+++ b/pppd/plugins/radius/radiusclient.h
@@ -440,6 +440,7 @@ UINT4 rc_get_ipaddr __P((char *));
 int rc_good_ipaddr __P((char *));
 const char *rc_ip_hostname __P((UINT4));
 UINT4 rc_own_ipaddress __P((void));
+UINT4 rc_own_bind_ipaddress __P((void));
 
 
 /*	sendserver.c		*/

--- a/pppd/plugins/radius/radrealms.c
+++ b/pppd/plugins/radius/radrealms.c
@@ -68,10 +68,12 @@ lookup_realm(char const *user,
     
     if ((fd = fopen(radrealms_config, "r")) == NULL) {
 	option_error("cannot open %s", radrealms_config);
+	free(auths);
+	free(accts);
 	return;
-    } 
+    }
     info("Reading %s", radrealms_config);
-    
+
     while ((fgets(buffer, sizeof(buffer), fd) != NULL)) {
 	line++;
 
@@ -87,6 +89,8 @@ lookup_realm(char const *user,
 	    fclose(fd);
 	    option_error("%s: invalid line %d: %s", radrealms_config,
 			 line, buffer);
+	    free(auths);
+	    free(accts);
 	    return;
 	}
 	info("Parsing '%s' entry:", p);
@@ -101,6 +105,8 @@ lookup_realm(char const *user,
 	    fclose(fd);
 	    option_error("%s: realm name missing on line %d: %s",
 			 radrealms_config, line, buffer);
+	    free(auths);
+	    free(accts);
 	    return;
 	}
 
@@ -111,6 +117,8 @@ lookup_realm(char const *user,
 		fclose(fd);
 		option_error("%s: server address missing on line %d: %s",
 			     radrealms_config, line, buffer);
+	        free(auths);
+	        free(accts);
 		return;
 	    }
 	    s->name[s->max] = strdup(p);
@@ -119,6 +127,8 @@ lookup_realm(char const *user,
 		fclose(fd);
 		option_error("%s: server port missing on line %d:  %s",
 			     radrealms_config, line, buffer);
+		free(auths);
+		free(accts);
 		return;
 	    }
 	    s->port[s->max] = atoi(p);

--- a/pppd/plugins/rp-pppoe/Makefile.linux
+++ b/pppd/plugins/rp-pppoe/Makefile.linux
@@ -33,10 +33,10 @@ pppoe-discovery: pppoe-discovery.o debug.o
 	$(CC) $(LDFLAGS) -o pppoe-discovery pppoe-discovery.o debug.o
 
 pppoe-discovery.o: pppoe-discovery.c
-	$(CC) $(CFLAGS) -c -o pppoe-discovery.o pppoe-discovery.c
+	$(CC) $(CFLAGS) -I../../.. -c -o pppoe-discovery.o pppoe-discovery.c
 
 debug.o: debug.c
-	$(CC) $(CFLAGS) -c -o debug.o debug.c
+	$(CC) $(CFLAGS) -I../../.. -c -o debug.o debug.c
 
 rp-pppoe.so: plugin.o discovery.o if.o common.o
 	$(CC) $(LDFLAGS) -o rp-pppoe.so -shared plugin.o discovery.o if.o common.o

--- a/pppd/plugins/rp-pppoe/if.c
+++ b/pppd/plugins/rp-pppoe/if.c
@@ -133,7 +133,7 @@ openInterface(char const *ifname, UINT16_t type, unsigned char *hwaddr)
 
     /* Fill in hardware address */
     if (hwaddr) {
-	strncpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
+	strlcpy(ifr.ifr_name, ifname, IFNAMSIZ);
 	if (ioctl(fd, SIOCGIFHWADDR, &ifr) < 0) {
 	    error("Can't get hardware address for %s: %m", ifname);
 	    close(fd);
@@ -152,7 +152,7 @@ openInterface(char const *ifname, UINT16_t type, unsigned char *hwaddr)
     }
 
     /* Sanity check on MTU */
-    strncpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
+    strlcpy(ifr.ifr_name, ifname, IFNAMSIZ);
     if (ioctl(fd, SIOCGIFMTU, &ifr) < 0) {
 	error("Can't get MTU for %s: %m", ifname);
     } else if (ifr.ifr_mtu < ETH_DATA_LEN) {
@@ -166,7 +166,7 @@ openInterface(char const *ifname, UINT16_t type, unsigned char *hwaddr)
     sa.sll_family = AF_PACKET;
     sa.sll_protocol = htons(type);
 
-    strncpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
+    strlcpy(ifr.ifr_name, ifname, IFNAMSIZ);
     if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0) {
 	error("Could not get interface index for %s: %m", ifname);
 	close(fd);

--- a/pppd/plugins/rp-pppoe/plugin.c
+++ b/pppd/plugins/rp-pppoe/plugin.c
@@ -153,7 +153,7 @@ PPPOEConnectDevice(void)
 	error("Can't get MTU for %s: %m", conn->ifName);
 	goto errout;
     }
-    strncpy(ifr.ifr_name, conn->ifName, sizeof(ifr.ifr_name));
+    strlcpy(ifr.ifr_name, conn->ifName, sizeof(ifr.ifr_name));
     if (ioctl(s, SIOCGIFMTU, &ifr) < 0) {
 	error("Can't get MTU for %s: %m", conn->ifName);
 	close(s);
@@ -327,7 +327,7 @@ PPPoEDevnameHook(char *cmd, char **argv, int doit)
 
     /* Try getting interface index */
     if (r) {
-	strncpy(ifr.ifr_name, cmd, sizeof(ifr.ifr_name));
+	strlcpy(ifr.ifr_name, cmd, sizeof(ifr.ifr_name));
 	if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0) {
 	    r = 0;
 	} else {
@@ -346,7 +346,7 @@ PPPoEDevnameHook(char *cmd, char **argv, int doit)
     /* Close socket */
     close(fd);
     if (r && doit) {
-	strncpy(devnam, cmd, sizeof(devnam));
+	strlcpy(devnam, cmd, sizeof(devnam));
 	if (the_channel != &pppoe_channel) {
 
 	    the_channel = &pppoe_channel;

--- a/pppd/plugins/rp-pppoe/pppoe-discovery.c
+++ b/pppd/plugins/rp-pppoe/pppoe-discovery.c
@@ -181,7 +181,8 @@ openInterface(char const *ifname, UINT16_t type, unsigned char *hwaddr)
     sa.sll_family = AF_PACKET;
     sa.sll_protocol = htons(type);
 
-    strncpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
+    strncpy(ifr.ifr_name, ifname, IFNAMSIZ);
+    ifr.ifr_name[IFNAMSIZ - 1] = 0;
     if (ioctl(fd, SIOCGIFINDEX, &ifr) < 0) {
 	fatalSys("ioctl(SIOCFIGINDEX): Could not get interface index");
     }

--- a/pppd/plugins/rp-pppoe/pppoe.h
+++ b/pppd/plugins/rp-pppoe/pppoe.h
@@ -22,6 +22,8 @@
 #include <stdio.h>		/* For FILE */
 #include <sys/types.h>		/* For pid_t */
 
+#include "pppd/pppd.h"		/* For error */
+
 /* How do we access raw Ethernet devices? */
 #undef USE_LINUX_PACKET
 #undef USE_BPF

--- a/pppd/plugins/winbind.c
+++ b/pppd/plugins/winbind.c
@@ -432,6 +432,7 @@ unsigned int run_ntlm_auth(const char *username,
 
         /* parent */
         if (close(child_out[0]) == -1) {
+                close(child_in[1]);
                 notice("error closing pipe?!? for child OUT[0]");
                 return NOT_AUTHENTICATED;
         }

--- a/pppd/sys-linux.c
+++ b/pppd/sys-linux.c
@@ -2167,7 +2167,6 @@ int ppp_available(void)
 		}
 	    }
 
-	    close (s);
 	    if (!ok) {
 		slprintf(route_buffer, sizeof(route_buffer),
 			 "Sorry - PPP driver version %d.%d.%d is out of date\n",
@@ -2177,6 +2176,7 @@ int ppp_available(void)
 	    }
 	}
     }
+    close(s);
     return ok;
 }
 
@@ -2653,7 +2653,10 @@ get_pty(master_fdp, slave_fdp, slave_name, uid)
 		warn("Couldn't unlock pty slave %s: %m", pty_name);
 #endif
 	    if ((sfd = open(pty_name, O_RDWR | O_NOCTTY)) < 0)
+	    {
 		warn("Couldn't open pty slave %s: %m", pty_name);
+		close(mfd);
+	    }
 	}
     }
 #endif /* TIOCGPTN */

--- a/pppstats/pppstats.c
+++ b/pppstats/pppstats.c
@@ -149,7 +149,8 @@ get_ppp_stats(curp)
 #define ifr_name ifr__name
 #endif
 
-    strncpy(req.ifr_name, interface, sizeof(req.ifr_name));
+    strncpy(req.ifr_name, interface, IFNAMSIZ);
+    req.ifr_name[IFNAMSIZ - 1] = 0;
     if (ioctl(s, SIOCGPPPSTATS, &req) < 0) {
 	fprintf(stderr, "%s: ", progname);
 	if (errno == ENOTTY)
@@ -175,7 +176,8 @@ get_ppp_cstats(csp)
 #define ifr_name ifr__name
 #endif
 
-    strncpy(creq.ifr_name, interface, sizeof(creq.ifr_name));
+    strncpy(creq.ifr_name, interface, IFNAMSIZ);
+    creq.ifr_name[IFNAMSIZ - 1] = 0;
     if (ioctl(s, SIOCGPPPCSTATS, &creq) < 0) {
 	fprintf(stderr, "%s: ", progname);
 	if (errno == ENOTTY) {
@@ -521,7 +523,8 @@ main(argc, argv)
 #undef  ifr_name
 #define ifr_name ifr_ifrn.ifrn_name
 #endif
-	strncpy(ifr.ifr_name, interface, sizeof(ifr.ifr_name));
+	strncpy(ifr.ifr_name, interface, IFNAMSIZ);
+	ifr.ifr_name[IFNAMSIZ - 1] = 0;
 	if (ioctl(s, SIOCGIFFLAGS, (caddr_t)&ifr) < 0) {
 	    fprintf(stderr, "%s: nonexistent interface '%s' specified\n",
 		    progname, interface);

